src/Stratis.Bitcoin/P2P/PeerConnector.cs

```diff
@@ -140,21 +155,9 @@ protected PeerConnector(NodeSettings nodeSettings, IPeerAddressManager peerAddre
         }
 
         /// <inheritdoc/>
-        public void Initialize(IConnectionManager connectionManager)
+        public void Initialize(NetworkPeerConnectionParameters parameters, Func<NetworkPeerConnectionParameters, NetworkPeerConnectionParameters> cloneParameters)
         {
-            this.asyncLoopFactory = connectionManager.AsyncLoopFactory;
-            this.ConnectedPeers = new NetworkPeerCollection();
-            this.loggerFactory = connectionManager.LoggerFactory;
-            this.network = connectionManager.Network;
-            this.networkPeerFactory = connectionManager.NetworkPeerFactory;
-            this.nodeLifetime = connectionManager.NodeLifetime;
-            this.NodeSettings = connectionManager.NodeSettings;
-            this.peerAddressManager = connectionManager.PeerAddressManager;
-
-            this.parentParameters = connectionManager.Parameters;
-            this.CurrentParameters = this.parentParameters.Clone();
-            this.CurrentParameters.ConnectCancellation = this.nodeLifetime.ApplicationStopping;
-            this.CurrentParameters.TemplateBehaviors.Add(new ConnectionManagerBehavior(false, connectionManager, this.loggerFactory));
+            this.CurrentParameters = cloneParameters(parameters);
```

Comment:
This must be a nightmare to test.
I don't see the point of the delegate, the same delegate code is used on all connectors so you could just put the code here instead of calling a delegate. 

What more is there we already use inheritance so if anything the logic could go inside  `OnInitialize()` on each child.

Reply:
You could also just have that logic in an extension method

Reply:
@dangershony Fair enough 👍 I wanted to get away from passing connection manager to the PeerConnector... I have realized now that the delegate is a bit over kill. I could have just cloned the parameters in a method and added the connection manager behavior to it. Then just pass those parameters to the connector.